src/main/scala/com/fulcrumgenomics/commons/async/AsyncRunnable.scala

```diff
@@ -0,0 +1,168 @@
+  * waiting.
+  */
+trait AsyncRunnable extends Runnable {
+
+  /** A reference to an exception thrown during an asynchronous operation. */
+  private[AsyncRunnable] val _throwable: AtomicReference[Throwable] = new AtomicReference(null)
+
+  /** This signals that the run method has started. */
+  private val startedLatch: CountDownLatch = new CountDownLatch(1)
+
+  /** This signals that the run method has completed, even if an exception occurred. This is needed so the close method
+    * can be sure that the run method is no longer executing. */
+  private val doneLatch: CountDownLatch = new CountDownLatch(1)
+
+  final def run(): Unit = {
+    val outer = this
+    startedLatch.countDown()
+    try {
+      outer.execute()
+    } catch {
+      case thr: Throwable =>
+        outer._throwable.compareAndSet(null, thr)
+        outer.uponException()
+    } finally {
+      outer.uponFinally()
+      this.doneLatch.countDown()
+    }
+  }
+
+  /** Returns true if the [[run()]] method has started, false otherwise. */
+  def started: Boolean = this.startedLatch.getCount == 0
+
+  /** Returns true if the [[run()]] method has completed, false otherwise. */
+  def done: Boolean = this.doneLatch.getCount == 0
+
+  /** Returns a throwable if an exception occurred in the [[run()]] method, None otherwise. */
+  def throwable: Option[Throwable] = Option(this._throwable.get())
+
+  /** Waits for the [[run()]] method to start. */
+  def awaitStart(): Unit = this.startedLatch.await()
+
+  /** Returns an [[Awaitable]] that completes when the [[run()]] method has started.  Returns the throwable if an
+    * exception was encountered, [[None]] otherwise. */
+  def startAwaitable()(implicit ec: ExecutionContext) : Awaitable[Unit] = Future { this.awaitStart() }
+
+  /** Waits for the [[run()]] method to complete. */
+  def awaitDone(): Unit = this.doneLatch.await()
+
+  /** Returns an [[Awaitable]] that completes when the [[run()]] method has completed.  Returns the throwable if an
+    * exception was encountered, [[None]] otherwise. */
+  def doneAwaitable()(implicit ec: ExecutionContext) : Awaitable[Option[Throwable]] = Future { this.awaitDone(); Option(this._throwable.get()) }
+
+  /** Creates a new thread wrapping this runnable; the thread is not started.
+    *
+    * @param name optionally the name of the thread, otherwise a name is created based on the class name and the number
+    *             of threads already created.
+    */
+  final def thread(name: Option[String] = None): Thread = {
+    require(!this.started, "Already started.")
+    val threadName = name.getOrElse { AsyncRunnable.nextName(this) }
+    new Thread(this, threadName)
+  }
+
+  /** Starts this [[Runnable]] in a daemon thread.
+    *
+    * @param name optionally the name of the thread, otherwise a name is created based on the class name and the number
+    *             of threads already created.
+    */
+  final def start(name: Option[String] = None): this.type = {
```

Comment:
How about making `daemon: Boolean = true` a parameter?

Reply:
Also, how about making `name` an instance method that defaults to `AsyncRunnable.name(this)`?  That way naming is also encapsulated in the class and not delegated to the user (who can still extend/override if they need to)?

Reply:
I think name would have to be a `val` since if it is a `def` then each call to name would return a different value based on how `AsyncRunnable.nextName` works.  Done.